### ISSUE TITLE
Rendering problem with stack exchange zim files

### ISSUE THREAD

**Issue by kelson42** [kiwix/apple#111]
Here is the online rendering:
http://library.kiwix.org/dsp.stackexchange.com_en_all_2017-10/A/question/374/river-detection-in-text.html

Here is how it is rendered on my iPad.
![934b210f-ed29-4e7d-862a-5ebde4d26b42](https://user-images.githubusercontent.com/1029718/36325481-34aee87a-1358-11e8-9d05-c9bf7b31cb20.png)

**Comment by automactic** [comment]
The reason this happens is because of relative `href` in html. I do not know what zim file you are using, but I have noticed this happening to stack exchange zims. So I randomly downloaded one and investigated.

Here is what happened:
1. I loaded a article of URL `kiwix://022efd27-391a-91f4-9f4e-2f1e53d41c94/A/question/6023.html`

**Comment by kelson42** [comment]
@automactic You can usually see the filename (without the trailing .zim") in kiwix-serve URLS. Here it is `dsp.stackexchange.com_en_all_2017-10.zim`.

We need to work on the same example, other this will be quite complicated.

Here is a bit explained what goes wrong here:
```
$ zimdump -f "question/374/river-detection-in-text.html" -i /var/www/download.kiwix.org/zim/stack_exchange/dsp.stackexchange.com_en_all_2017-10.zim
url: question/374/river-detection-in-text.html
	title:           Question 374
	idx:             35470
	namespace:       A
	type:            redirect
	redirect index:  35469
$ zimdump -f "question/374.html" -i /var/www/download.kiwix.org/zim/stack_exchange/dsp.stackexchange.com_en_all_2017-10.zim
url: question/374.html
	title:           "River" detection in text
	idx:             35469
	namespace:       A
	type:            article
	mime-type:       text/html
	article size:    32740
```

The root cause is here:
https://github.com/kiwix/kiwix-lib/issues/70

**Comment by kelson42** [comment]
@automactic Would you be able to deal with the "base" value returned by the kiwix-lib... at least as long as we do not return the redirect content and then you will have to handle the redirection process on your own.

**Comment by automactic** [comment]
@kelson42, I do not understand what you said there.

> Would you be able to deal with the "base" value returned by the kiwix-lib
Is this supposed to be a question??
What "base" value?

How should I figure our what is the redirected URL (if there's one)?

**Comment by kelson42** [comment]
@automactic Do know more about the the base (url) value, read this comment by Matthieu https://github.com/kiwix/kiwix-lib/issues/70#issuecomment-326923138. For the rest you might need to read that documentation https://www.w3schools.com/tags/tag_base.asp

**Comment by automactic** [comment]
@kelson42, in your opinion, is this a issue where the app didn’t render the html correctly, or the html provided by the zim file is incorrect, or libkiwix hasn’t provided the function to allow the html to be rendered correctly?

**Comment by automactic** [comment]
I personally think there is nothing to fix in the app. The app loads a webpage, the html of the webpage have resources that are not there, so some of the resource loading failed. 

@kelson42, please tell me if you think there is something in the app that should be fixed? Or if there is a way to check if a url should be redirected? If not, please close this issue.

**Comment by kelson42** [comment]
@automactic It is a problem between the app (the URL asked) and the kiwixlib (the content delivered) - both are ok, but not together as basically the URL of the content delivered is an other one (because of the redirect).

We have now a solution provided which would allow you to fix the problem, like kiwix-serve does, by adding a `<base>` node... but the ultimate solution is on the libkiwix side, which indeed should give you the redirect target URL, so you can handle the redirection yourself (in the app).

**Comment by automactic** [comment]
@kelson42, just to be clear. Who is supposed to add `<base>`? The kiwix for iOS app or the zim file?

**Comment by kelson42** [comment]
@automactic you, you can have a look to the kiwix-serve source how we did it. That way it will fix the problem and give us the few months we need to tacle the problem correctly.

**Comment by automactic** [comment]
I haven't read kiwix-serve source code (also there doesn't seems to be a kiwix-serve repo). But I hope you are not suggesting we manually parse the html string and add a base tag. Because if it is, I would be very reluctant to do so. 

Kiwix app is supposed to be used as a generic zim file reader, we should not have file specific routines hardcoded in the app. Especially when there is no definite way of testing if such fix should be applied. It kind feels like someone ask nginx to modify their code when his site is not working, asking vlc to fix their decoded when a video file cannot be played. On the other hand, if we are releasing a dedicated stack exchange offline app like wikimed, I would be very happily add this fix, since we are dealing with one zim file only. 

I personally think the stack exchange zim files clearly have problems, or maybe the problem is with redirection.  If it is not ready yet, don't release them. Wait till they are ready and release with confidence.

**Comment by kelson42** [comment]
@automactic What would you imagine for a kiwix-lib behaviour if a content is a redirect?

**Comment by kelson42** [comment]
Fixed in 1.9 (testflight 12)
